### ISSUE TITLE
[question] cannot recover the optimal reward from saved best_model.zip as the tensorboard reported

### ISSUE THREAD

**Issue by xibeisiber** [DLR-RM/rl-baselines3-zoo#282]
Hi, 
I'm using the train.py and enjoy.py to train my own environments.

`python train.py --algo sac --env BounceBallPyBullet-v0 --verbose 1 --eval-episodes 5 --n-eval-envs 5`
`python enjoy.py --algo sac --env BounceBallPyBullet-v0 -f logs/ --exp-id 5 --load-best --env-kwargs gym_render:True`

The default value of "deterministic_eval" should be true for my env. After the training, I saw from the tensorboard the "eval/mean_reward" has a highest value 1082, "eval/mean_ep_length" reached the max_episode_steps. 

But when I use enjoy.py to load the best_model, the eval episode reward and length are very low. I checked the timestamp of the best_model.zip, it is modified at the same time of the highest value in the tensorboard. Since deterministic_eval=True, I don't understand why I cann't recover the reward.

Also, when I modify the exp_manager.py by setting the self.deterministic_eval to False, in the training process I print the "episode_rewards, episode_lengths" defined in line 393 in stable-baselines3/common/callbacks.py. The episode_rewards and episode_lengths each have 5 same values with std=0. I expected there should be 5 different values since deterministic_eval is False.

BTW: The hyperparameters are not fully tuned. Since in the env I set done=True when something undesired happened, if max_episode_steps (30000) is reached, this would mean the policy is quite good. And sometimes in other training cases I use enjoy.py with the best_model.zip corresponding to the peaks in the eval/mean_ep_length, it indeed exhibits very good results. 

![eval_mean_ep_length](https://user-images.githubusercontent.com/5679317/187652254-e9dbefd0-97ce-40f9-aefd-28bdddf344e5.png)

![eval_mean_reward](https://user-images.githubusercontent.com/5679317/187652287-8bc3dd96-057c-4039-9ab5-dafa8f00720d.png)

![rollout_ep_len_mean](https://user-images.githubusercontent.com/5679317/187652320-2ad413c7-6e41-4b59-a751-752ff8e27b24.png)

![rollout_ep_rew_mean](https://user-images.githubusercontent.com/5679317/187652364-326c0dab-7595-472d-b7f5-8224c49b9ce6.png)

Thanks for any help!

**Comment by araffin** [comment]
Replying from https://github.com/DLR-RM/stable-baselines3/issues/1045

> The default value for "deterministic_eval = not self.is_atari(env_id)" in exp_manager.py is True for my own custom env. So >evaluating for more episodes will give the same result.

>But if I set "deterministic_eval=False" manually, evaluating for 5 episodes always gives the same results. I assume something is >missing for my env....


There is probably a misunderstanding of what `deterministic_eval` means.
It means that the deterministic controller will used for evaluation (not sampling action for the action distribution but using the mode of the distribution, think "greedy" in the case of discrete actions) but if the initial conditions are different (initial observation) or the env is stochastic, then you will get different results for different episodes.

> evaluating for 5 episodes always gives the same results. I assume something is >missing for my env....

yes, double check your env.

**Comment by xibeisiber** [comment]
I may find the possible reason that the saved best model has a quite different reward when evaluated using enjoy.py compared to the reward in tensorboard curve.

My env is totally deterministic, and I **set "norm_obs" to True** when training. The training script will save a **vecnormalize.pkl** when a new best_model is saved. But when the training process ends, it will again save a vecnormalize.pkl in the function save_trained_model(). This last pkl will overwrite the pkl corresponding to the best model. Therefore when enjoying the best_model, it will load the last pkl and give a very different reward if the statistics in the two pkls differ too much.

**Comment by araffin** [comment]
> l. Therefore when enjoying the best_model, it will load the last pkl and give a very different reward if the statistics in the two pkls differ too much.

Related to https://github.com/DLR-RM/rl-baselines3-zoo/issues/278

Normally the stats should not differ too much once it's converged.
To double check that's actually the problem, you can set `training=True` at test time when using `enjoy.py`.

With latest SB3 version, it should be easy to activate saving stats for every checkpoints: https://github.com/DLR-RM/stable-baselines3/issues/1016

**Comment by xibeisiber** [comment]
> Related to #278

Thanks for your reply! I forked and modify the rl-baselines3-zoo/utils/callback.py to save multi best_models and at the same time rename each vecnormalize.pkl to match the id of best models. And modify the enjoy.py to load best_modes4.zip and vecnormalize4.pkl for evaluation. 

forked version is [here](https://github.com/xibeisiber/rl-baselines3-zoo/commit/2690615528cac970110d86f7a3a24a0910d3f794)

when the total training timesteps is small, the reward of each best_model coheres with the value in tensorboard
but when the total timesteps increase furthe, they are inconsistent again...... (*_ _)
